src/test/regress/sql/union_pushdown.sql

```diff
@@ -230,5 +236,381 @@ SELECT DISTINCT user_id FROM
 USING (user_id)
 ORDER BY 1 LIMIT 1;
 
+-- UNION ALL leaf queries deep in the subquery
+SELECT * FROM
+(
+	SELECT * FROM (SELECT * FROM (SELECT * FROM (SELECT * FROM (SELECT * FROM (SELECT * FROM users_table_part) as level_5) as level_4) as level_3) as level_2) as level_1
+	UNION ALL
+	SELECT * FROM (SELECT * FROM (SELECT * FROM (SELECT * FROM (SELECT * FROM (SELECT * FROM users_table_part) as level_5) as level_4) as level_3) as level_2) as level_1
+) as top_level ORDER BY 1 DESC LIMIT 3;
+
+-- UNION ALL leaf queries deep in the subquery
+-- and random() calls prevent any pullup
+SELECT user_id FROM
+(
+	SELECT *,random() FROM (SELECT *,random() FROM (SELECT *,random() FROM (SELECT *,random() FROM (SELECT *,random() FROM (SELECT *,random() FROM users_table_part) as level_5) as level_4) as level_3) as level_2) as level_1
+	UNION ALL
+	SELECT *,random() FROM (SELECT *,random() FROM (SELECT *,random() FROM (SELECT *,random() FROM (SELECT *,random() FROM (SELECT *,random() FROM users_table_part) as level_5) as level_4) as level_3) as level_2) as level_1
+) as top_level ORDER BY 1 DESC LIMIT 3;
+
+
+-- UNION ALL leaf queries deep in the subquery
+-- joined with a table
+SELECT * FROM
+(
+	SELECT * FROM (SELECT * FROM (SELECT * FROM (SELECT * FROM (SELECT * FROM (SELECT * FROM users_table_part UNION ALL SELECT * FROM users_table_part) as level_5) as level_4) as level_3) as level_2) as level_1
+	UNION ALL
+	SELECT * FROM (SELECT * FROM (SELECT * FROM (SELECT * FROM (SELECT * FROM (SELECT * FROM users_table_part UNION ALL SELECT * FROM users_table_part) as level_5) as level_4) as level_3) as level_2) as level_1
+) as top_level
+JOIN
+events_table_part USING(user_id)
+ ORDER BY 1 DESC LIMIT 3;
+
+-- UNION ALL leaf queries deep in the subquery
+-- and random() calls prevent any pullup
+SELECT user_id FROM
```

Comment:
Minor: is there any value in having union all that is not in leaf or at top level but in between?

Reply:
I think the main distinction is being top-level vs not top level in the code, and we have quite a lot of different cases in the test now that triggers various queryTree structures